keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+Kubernetes REST API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Configuring authentication via command line flags has some limitations.  Some
+forms of config are simply easier to specify and understand in a structured
+manner such as via a Kubernetes API or file.  The existing authentication flags
+for OIDC and token webhook limit the user to only one of each of these types of
+authentication modes as there is no way to specify the set of flags multiple
+times [#71162].  These flag based configs also require an API server restart to
+take effect.
+
+Since there is no way to configure authentication other than via flags, hosted
+environments have no consistent way to expose this functionality [#166].  To get
+around this limitation, users install authentication proxies that run as
+cluster-admin and impersonate the desired user [kube-oidc-proxy] [teleport].
+Other than being a gross abuse of the impersonation API, this opens the API
+server to escalation bugs caused by the proxy (such as improper handling of
+incoming request headers).  This proxy also intercepts network traffic which
+poses privacy and scalability concerns.
+
+As a simple case study, it should be possible to deploy a simple identity
+provider (ex: [dex]) directly onto a running cluster and be able to use that IDP
+to authenticate users in the order of minutes without restarting the API server.
+No proxying of network traffic should occur and the impersonation API must not
+be used.
+
+[#71162]: https://github.com/kubernetes/kubernetes/issues/71162
+[#166]: https://github.com/aws/containers-roadmap/issues/166
+[kube-oidc-proxy]: https://aws.amazon.com/blogs/opensource/consistent-oidc-authentication-across-multiple-eks-clusters-using-kube-oidc-proxy
+[teleport]: https://aws.amazon.com/blogs/opensource/authenticating-eks-github-credentials-teleport
+[dex]: https://github.com/dexidp/dex
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+1. Create a new Kubernetes REST API that allows configuration of authentication
+  - x509
+  - OIDC
+  - Token Webhook
+2. Changes made via the REST API should be active in the order of minutes without
+requiring a restart of the API server
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+1. Changing the internal behavior of the authentication modes (ex: using
+different fields in x509 certificates to build the `user.Info` object)
+2. Restricting the REST API to the exact structure of the CLI flags
```

Comment:
What about preservation of the existing authn flags? I can see paranoid admins who would be afraid of allowing dynamic authn and prefer not to use the API.

Reply:
> What about preservation of the existing authn flags?

Not sure what you mean.  I am not proposing we change the CLI flags in any way.

> I can see paranoid admins who would be afraid of allowing dynamic authn and prefer not to use the API.

I would assume the paranoid admin would not give users cluster-admin access?

Reply:
Something to consider, if we gave this API its own group, then we could use `--runtime-config` to allow disabling of that single API.  That might be overkill / an undesired approach.

A better approach may be to allow configuring of admission webhooks via CLI flags so that an infrastructure provider could guard some subset of admission plugins from the cluster-admin (presumably that cluster-admin would not have direct access to the API server host).

Something we could do today is include an admission plugin that is disabled by default that, once enabled, would reject any `create` and `delete` requests (unless `deletionTimestamp` was set) to this API and only allow `updates` to via `IsOnlyMutatingGCFields`.

Reply:
> I just mean I wouldn't want to see existing CLI flags deprecated in favor of this.

Yeah, I do not see us changing the CLI functionality in any way since the CLI flags allow for static, unchangeable config.  IMO there is enough value in the CLI flags that even with this new functionality, we would not attempt to deprecate the flags.

Reply:
@enj to resolve this question, do you want to add a new non-goal of deprecating or removing the existing flags?